libkbfs/block_types.go

```diff
@@ -186,12 +188,51 @@ func (fb *FileBlock) NewEmpty() Block {
 	return &FileBlock{}
 }
 
-// DataVersion returns data version for this block.
+// DataVersion returns data version for this block, which is assumed
+// to have been modified locally.
 func (fb *FileBlock) DataVersion() DataVer {
+	if !fb.IsInd {
+		return FirstValidDataVer
+	}
+
+	// If this is an indirect block, and none of its children are
+	// marked as direct blocks, then this must be a big file.  Note
+	// that we do it this way, rather than returning on the first
+	// non-direct block, to support appending to existing files and
+	// making them big.
+	hasHoles := false
+	hasDirect := false
+	maxDirectType := UnknownDirectType
 	for i := range fb.IPtrs {
-		if fb.IPtrs[i].Holes {
-			return FilesWithHolesDataVer
+		if maxDirectType != UnknownDirectType &&
+			fb.IPtrs[i].DirectType != UnknownDirectType &&
+			maxDirectType != fb.IPtrs[i].DirectType {
+			panic("Mixed data versions among indirect pointers")
+		}
+		if fb.IPtrs[i].DirectType > maxDirectType {
+			maxDirectType = fb.IPtrs[i].DirectType
+		}
+
+		if fb.IPtrs[i].DirectType == DirectBlock {
+			hasDirect = true
+		} else if fb.IPtrs[i].Holes {
+			hasHoles = true
 		}
+		// We can only safely break if both vars are definitely set to
+		// their final value.
+		if hasDirect && hasHoles {
+			break
+		}
+	}
+
+	if maxDirectType == UnknownDirectType {
+		panic("No known type for any indirect pointer")
+	}
+
+	if len(fb.IPtrs) > 0 && !hasDirect {
```

Comment:
the `len` check is redundant here (since if `len == 0`, the above panic would have hit)